packages/manager/apps/catalog-revamp/src/utils/utils.ts

```diff
@@ -0,0 +1,140 @@
+};
+
+type ProductKey = 'universe' | 'category' | 'name';
+
+export const groupItemsByUniverse = (itemsToGroup: Product[]) => {
+  return itemsToGroup.reduce<Record<string, Product[]>>((acc, item) => {
+    if (!acc[item.universe]) acc[item.universe] = [];
+    acc[item.universe].push(item);
+    return acc;
+  }, {});
+};
+
+export const filterByEnum = (
+  product: Product,
+  query: string[],
+  property: ProductKey,
+) => {
+  return query.length === 0 || query.includes(product[property]);
+};
+
+export const getAvailableCategories = (
+  products: Product[],
+  selectedUniverses: string[],
+) => {
+  const itemsFiltered = products.filter((product) =>
+    filterByEnum(product, selectedUniverses, 'universe'),
+  );
+  const uniqueCategories = [
+    ...new Set(itemsFiltered.map((item) => item.category)),
+  ];
+  return uniqueCategories.sort((a, b) =>
+    a.localeCompare(b, undefined, { sensitivity: 'base' }),
+  );
+};
+
+export const matchCategories = (
+  product: Product,
+  selectedCategories: string[],
+) => {
+  return filterByEnum(product, selectedCategories, 'category');
+};
+
+export const matchUniverses = (
+  product: Product,
+  selectedUniverses: string[],
+) => {
+  return filterByEnum(product, selectedUniverses, 'universe');
+};
+
+export const matchSearchText = (product: Product, searchText: string) => {
+  const search = searchText?.toLowerCase();
+  return (
+    !search ||
+    product.category.toLowerCase().includes(search) ||
+    product.universe.toLowerCase().includes(search) ||
+    product.name.toLowerCase().includes(search)
+  );
+};
+
+export const filterProducts = (
+  products: Product[],
+  selectedCategories: string[],
+  selectedUniverses: string[],
+  searchText: string,
+) => {
+  return products.filter(
+    (product) =>
+      matchCategories(product, selectedCategories) &&
+      matchUniverses(product, selectedUniverses) &&
+      matchSearchText(product, searchText),
+  );
+};
+
+// getUniverses overloads
+export function getUniverses(
+  products: Product[],
+  withCounter: true,
+): { universe: string; count: number }[];
+export function getUniverses(
+  products: Product[],
+  withCounter?: false,
+): string[];
+// getUniverses implementation
+export function getUniverses(products: Product[], withCounter = false): any {
+  if (withCounter) {
+    const universeCounts = products.reduce((acc, product) => {
+      acc[product.universe] = (acc[product.universe] || 0) + 1;
+      return acc;
+    }, {} as Record<string, number>);
+
+    const formattedUniverses = Object.entries(
+      universeCounts,
+    ).map(([universe, count]) => ({ universe, count }));
+
+    return formattedUniverses.sort((a, b) =>
+      a.universe.localeCompare(b.universe, undefined, { sensitivity: 'base' }),
+    );
+  }
+
+  const uniqueUniverses = [
+    ...new Set(products.map((product) => product.universe)),
+  ];
+  return uniqueUniverses.sort((a, b) =>
+    a.localeCompare(b, undefined, { sensitivity: 'base' }),
+  );
+}
```

Comment:
how is it better than creating 2 distinct functions ?

Reply:
No that's not a generic function, that's 2 different functions in one function and you use one or the other based on a parameter. There is nothing in a common between the 2 functions, except that they both take `products` as a parameter.

Just write 2 distinct functions `getUniversesWithCounter` and `getUniverseList`.